'array' variable name fails .map() and .filter() methods

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** 
V full version: V 0.2.2 f48f623.ae05ab8
**OS:** 
OS: macos, macOS, 11.5.1, 20G80
Processor: 8 cpus, 64bit, little endian, Apple M1
CC version: Apple clang version 12.0.5 (clang-1205.0.22.11)

**What did you do?**
Used a variable ```array``` and program fails to compile using ```array.map()``` and ```array.filter()```.  Other methods ```.reduce(), .any(), .all()``` all work as expected.  Changing the variable name fixes the problem with ```.map()``` and ```.filter()```.  Was mentioned in https://github.com/vlang/v/issues/10982
```v
//------------error----------------
	array := [1,2,3,4]
	mut res1 := array.map( it * 3 )
	mut res2 := array.filter( it > 2 )
	println(res1)
	println(res2)

//--------------ok-----------------
	// arr := [1,2,3,4]
	// mut res3 := arr.map( it * 3 )
	// mut res4 := arr.filter( it > 2 )
	// println(res3)
	// println(res4)
```

**What did you expect to see?**
```
[3, 6, 9, 12]
[3, 4]
```
 
**What did you see instead?**
```
/tmp/v/arrerror.6199580445162401129.tmp.c:10209:22: error: expected expression
                        array_push((array*)&_t3, &it); 
                                          ^
2 errors generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1495,8 +1495,10 @@ fn (mut g Gen) stmt(node ast.Stmt) {
 		ast.StructDecl {
 			name := if node.language == .c {
 				util.no_dots(node.name)
+			} else if node.name in ['array', 'string'] {
+				node.name
 			} else {
-				if node.name in ['array', 'string'] { node.name } else { c_name(node.name) }
+				c_name(node.name)
 			}
 			// TODO For some reason, build fails with autofree with this line
 			// as it's only informative, comment it for now

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -15,12 +15,12 @@ import v.depgraph
 const (
 	// NB: some of the words in c_reserved, are not reserved in C,
 	// but are in C++, or have special meaning in V, thus need escaping too.
-	c_reserved     = ['auto', 'break', 'calloc', 'case', 'char', 'class', 'const', 'continue',
-		'default', 'delete', 'do', 'double', 'else', 'enum', 'error', 'exit', 'export', 'extern',
-		'float', 'for', 'free', 'goto', 'if', 'inline', 'int', 'link', 'long', 'malloc', 'namespace',
-		'new', 'panic', 'register', 'restrict', 'return', 'short', 'signed', 'sizeof', 'static',
-		'struct', 'switch', 'typedef', 'typename', 'union', 'unix', 'unsigned', 'void', 'volatile',
-		'while', 'template', 'stdout', 'stdin', 'stderr']
+	c_reserved     = ['array', 'auto', 'break', 'calloc', 'case', 'char', 'class', 'const',
+		'continue', 'default', 'delete', 'do', 'double', 'else', 'enum', 'error', 'exit', 'export',
+		'extern', 'float', 'for', 'free', 'goto', 'if', 'inline', 'int', 'link', 'long', 'malloc',
+		'namespace', 'new', 'panic', 'register', 'restrict', 'return', 'short', 'signed', 'sizeof',
+		'static', 'string', 'struct', 'switch', 'typedef', 'typename', 'union', 'unix', 'unsigned',
+		'void', 'volatile', 'while', 'template', 'stdout', 'stdin', 'stderr']
 	c_reserved_map = string_array_to_map(c_reserved)
 	// same order as in token.Kind
 	cmp_str        = ['eq', 'ne', 'gt', 'lt', 'ge', 'le']
@@ -1493,7 +1493,11 @@ fn (mut g Gen) stmt(node ast.Stmt) {
 			g.sql_stmt(node)
 		}
 		ast.StructDecl {
-			name := if node.language == .c { util.no_dots(node.name) } else { c_name(node.name) }
+			name := if node.language == .c {
+				util.no_dots(node.name)
+			} else {
+				if node.name in ['array', 'string'] { node.name } else { c_name(node.name) }
+			}
 			// TODO For some reason, build fails with autofree with this line
 			// as it's only informative, comment it for now
 			// g.gen_attrs(node.attrs)

--- a/vlib/v/tests/reserved_keywords_array_and_string_test.v
+++ b/vlib/v/tests/reserved_keywords_array_and_string_test.v
@@ -1,0 +1,13 @@
+fn test_reserved_keywords_array_and_string() {
+	array := [1, 2, 3, 4]
+	mut res1 := array.map(it * 3)
+	mut res2 := array.filter(it > 2)
+	println(res1)
+	assert res1 == [3, 6, 9, 12]
+	println(res2)
+	assert res2 == [3, 4]
+
+	string := 'hello'
+	println(string)
+	assert string == 'hello'
+}

